File: src/rate_limiter/rate_limiter.erl
-module(rate_limiter).

% elp:ignore W0012 W0040 E1599
-moduledoc {file, "../../docs/rate_limiter.md"}.

-export([
    start_worker/0,
    terminate_worker/0,
    get_worker_concurrency_limit/0,
    set_worker_concurrency_limit/1,
    start/0
]).

-define(WORKER_ID, "limited_worker").

-define(FREQUENCY_LIMIT, 1_000).

start_worker() ->
    temporal_sdk_worker:start(
        cluster_1,
        activity,
        [
            {worker_id, ?WORKER_ID},
            {task_queue, "limited_tq"},
            {task_poller_pool_size, 1}
        ]
    ).

terminate_worker() ->
    temporal_sdk_worker:terminate(cluster_1, activity, ?WORKER_ID).

get_worker_concurrency_limit() ->
    case temporal_sdk_worker:get_limiter_config(cluster_1, activity, ?WORKER_ID) of
        {ok, #{limits := #{worker := #{activity_regular := {L, _}}}}} -> {ok, L};
        {ok, _} -> undefined;
        Err -> Err
    end.

set_worker_concurrency_limit(ConcurrencyLimit) when is_integer(ConcurrencyLimit) ->
    temporal_sdk_worker:set_limiter_config(
        cluster_1,
        activity,
        ?WORKER_ID,
        #{limits => #{worker => #{activity_regular => {ConcurrencyLimit, ?FREQUENCY_LIMIT}}}}
    ).

start() ->
    StartTime = erlang:system_time(millisecond),
    case temporal_sdk:start_workflow(cluster_1, "default", rate_limiter_workflow, [wait]) of
        {#{}, {completed, #{}}} ->
            Duration = erlang:system_time(millisecond) - StartTime,
            io:fwrite("Workflow completed in ~b msec.~n", [Duration]);
        Err ->
            Err
    end.

File: lib/rate_limiter/rate_limiter.ex
defmodule RateLimiter do
  @external_resource "docs/rate_limiter.md"
  @moduledoc TemporalSdk.Utils.exdoc!("docs/rate_limiter.md")

  @worker_id "limited_worker"
  @frequency_limit 1_000

  def start_worker() do
    TemporalSdk.Worker.start(
      :cluster_1,
      :activity,
      worker_id: @worker_id,
      task_queue: "limited_tq",
      task_poller_pool_size: 1
    )
  end

  def terminate_worker(), do: TemporalSdk.Worker.terminate(:cluster_1, :activity, @worker_id)

  def get_worker_concurrency_limit() do
    case TemporalSdk.Worker.get_limiter_config(:cluster_1, :activity, @worker_id) do
      {:ok, %{limits: %{worker: %{activity_regular: {l, _}}}}} -> {:ok, l}
      {:ok, _} -> :undefined
      err -> err
    end
  end

  def set_worker_concurrency_limit(concurrency_limit) when is_integer(concurrency_limit) do
    TemporalSdk.Worker.set_limiter_config(
      :cluster_1,
      :activity,
      @worker_id,
      %{:limits => %{:worker => %{:activity_regular => {concurrency_limit, @frequency_limit}}}}
    )
  end

  def start() do
    start_time = System.system_time(:millisecond)

    case TemporalSdk.start_workflow(:cluster_1, "default", RateLimiter.Workflow, [:wait]) do
      {%{}, {:completed, %{}}} ->
        duration = System.system_time(:millisecond) - start_time
        IO.puts("Workflow completed in #{duration} msec.")

      err ->
        err
    end
  end
end
